### PR TITLE
circleci pipeline 

### PR DESCRIPTION
**Pull request by humulla**
- [x]  build and deploy job for dev, staging and production

- [x] add run test command for unit tests      

- [x]  add the S3 bucket sync once terraform for them have been created 

- [x]  add terraform job to pipeline

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    working_directory: *workspace_root
+    steps:
+      - checkout
+      - run:
+          name: Install dependencies
+          command: yarn install
+      - persist_to_workspace:
+          root: "~/repo"
+          paths: .
+           assume-role-development:
```
Comment (LBHMGeorgieva) [inline .circleci/config.yml:49]: Indentation needs to be fixed here

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+          aws-account: $AWS_ACCOUNT_STAGING
+
+  assume-role-production:
+    executor: fss-executor
+    steps:
+      - assume-role-and-persist-workspace:
+          aws-account: $AWS_ACCOUNT_PRODUCTION
+
+  build-deploy-development:
+    executor: fss-executor
```
Comment (LBHMGeorgieva) [inline .circleci/config.yml:68]: The executor name needs to be amended

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+          name: build
+          command: |
+            yarn run build
+      - run: 
+          name: test
+          command: |
+            yarn test
+          overwrite: true
+  build-deploy-staging:
+    executor: fss-executor
```
Comment (LBHMGeorgieva) [inline .circleci/config.yml:81]: Executor name needs to be changed

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+          name: build
+          command: |
+            yarn run build
+      - run: 
+          name: test
+          command: |
+            yarn test
+          overwrite: true
+  build-deploy-production:
+    executor: fss-executor
```
Comment (LBHMGeorgieva) [inline .circleci/config.yml:94]: Executor name needs to be changed

**Review by LBHMGeorgieva — CHANGES_REQUESTED** [review]
Please have a look at some of the comments I left. Also, the step to deploy to S3 is missing. I would suggest you separate the run tests command as a separate job, and in the build and deploy job, run the build command and deploy to S3. Lastly, please add a job for deploying the terraform resources.

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+      - *attach_workspace
+      - run: 
+          name: test
+          command: |
+            yarn test
+workflows:
+  version: 2
+  continuous-delivery:
+    jobs:
+      - install-dependencies
```
Comment (LBHMGeorgieva) [inline .circleci/config.yml:159]: We should run the tests after the dependencies are installed, but before we proceed with assuming the role and applying terraform

File: .circleci/config.yml [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+    executor: mtfh-executor
+    steps:
+      - *attach_workspace
+      - run:
+          name: build
+          command: |
+            yarn run build
+     - aws-s3/sync:
+          from: build
+          to: 's3://lbh-housing-tenancy-lease-frontend-development.hackney.gov.uk/'
```
Comment (LBHMGeorgieva) [inline .circleci/config.yml:127]: Bucket name needs to change to reflect environment

File: .circleci/config.yml [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+    executor: mtfh-executor
+    steps:
+      - *attach_workspace
+      - run:
+          name: build
+          command: |
+            yarn run build
+      - aws-s3/sync:
+          from: build
+          to: 's3://lbh-housing-tenancy-lease-frontend-development.hackney.gov.uk/'
```
Comment (LBHMGeorgieva) [inline .circleci/config.yml:142]: Bucket name needs to change to reflect environment

File: .circleci/config.yml (hunk 5 above)
Comment (humulla) [reply .circleci/config.yml:159]: `run tests` should require `build-deploy-development` to pass right? 
if so, would `build-deploy-development `be after `install-dependencies` and then `run-tests`?

File: .circleci/config.yml [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+          name: test
+          command: |
+            yarn test
+workflows:
+  version: 2
+  continuous-delivery:
+    jobs:
+      - install-dependencies
+      - run-tests:
+          context: api-assume-role-housing-development-context
```
Comment (LBHMGeorgieva) [inline .circleci/config.yml:158]: You don't need to pass the context here

File: .circleci/config.yml [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+          filters:
+            branches:
+              only:
+                - master
+      - assume-role-development:
+          context: api-assume-role-housing-development-context
+          requires:
+            - install-dependencies
```
Comment (LBHMGeorgieva) [inline .circleci/config.yml:166]: maybe change this to be requires run-tests

File: .circleci/config.yml [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+          requires:
+            - assume-role-development
+          filters:
+            branches:
+              only:
+                - master
+      - build-deploy-development:
+          context: api-assume-role-housing-development-context
+          requires:
+            - assume-role-development
```
Comment (LBHMGeorgieva) [inline .circleci/config.yml:175]: This should be requires :  terraform-init-and-apply-to-development

**Review by LBHMGeorgieva — APPROVED** [review]

**Pull request merged**